### PR TITLE
[FEATURE][ML] Progress monitoring

### PR DESCRIPTION
**Pull request by tveasey**
This wires in progress monitoring to the data frame analysis framework. I've discretised the progress so I can efficiently update atomically. The actual progress is updated via a callback to maximally decouple the "observer" from the algorithms whose progress is being observed. This also makes it trivial to stub out progress monitoring.

Note I haven't yet done anything to pass this back during a long running analysis, because the Java side isn't expecting anything yet.

### CODE REVIEW

File: lib/core/unittest/CLoopProgressTest.cc [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+#include <core/CLogger.h>
+#include <core/CLoopProgress.h>
+
+#include <test/CRandomNumbers.h>
+
+using namespace ml;
+
+using TSizeVec = std::vector<std::size_t>;
+
+void CLoopProgressTest::testShort() {
```
Comment (dimitris-athanasiou) [inline lib/core/unittest/CLoopProgressTest.cc]: I think it'd be be nice to add a simple test (say 4 steps) that also verifies the exact progress in after each increment.

File: lib/core/unittest/CLoopProgressTest.cc [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        LOG_DEBUG(<< "Testing " << n);
+
+        progress = 0.0;
+        core::CLoopProgress loopProgress{n,
+                                         [&progress](double p) { progress += p; }};
+
+        for (std::size_t i = 0; i < n; i += 2, loopProgress.increment(2)) {
+            // Empty
+        }
+    }
```
Comment (dimitris-athanasiou) [inline lib/core/unittest/CLoopProgressTest.cc]: missing assertion?

File: include/api/CDataFrameAnalysisRunner.h [hunk 3]
```diff
@@ -132,7 +132,17 @@ class API_EXPORT CDataFrameAnalysisRunner {
 
     void setToBad();
     void setToFinished();
-    void updateProgress(double fractionalProgress);
+
+    //! This adds \p fractionalProgess to the current progress.
+    //!
+    //! \note The caller should try to ensure that the sum of the values added
```
[the remark this answers is not in the document]
Comment (tveasey) [reply include/api/CDataFrameAnalysisRunner.h]: I think this is still valid. The overall progress monitoring might use multiple loop progress monitors with weights dependent on their proportion of the run time. For example, imagine we have at the top level, schematically

...
Loop A which takes 20% of the run time
...
Loop B which takes 80% of the run time
...

When the caller instruments this code for progress monitoring they have to scale the progress for loop A by w(A) = 0.2 and for loop B by w(B) = 0.8, for example by wrapping the callback recordProgress passed to the function invoking Loop A as follows `[&recordProgress](double p) { recordProgress(0.2 * p); }`.

It the responsibility of the caller to ensure that w(A) + w(B) = 1 to get decent progress monitoring, which is what this comment states.

File: lib/core/unittest/CLoopProgressTest.cc (hunk 2 above)
Comment (tveasey) [reply lib/core/unittest/CLoopProgressTest.cc]: I've added some incremental progress assertions: [f1d9ea1](https://github.com/elastic/ml-cpp/pull/362/commits/f1d9ea129087137db34347f269dd3c8e7242e3b3). These cover your other comment I think.

File: lib/core/unittest/CLoopProgressTest.cc (hunk 1 above)
Comment (tveasey) [reply lib/core/unittest/CLoopProgressTest.cc]: See below.

**Comment by tveasey** [comment]
Thanks @dimitris-athanasiou. I've addressed your comment on test coverage, can you take another look.

**Pull request merged**